redis/lookup_test.go

```diff
@@ -0,0 +1,224 @@
+}
+
+func TestNewLookupWithAllFailingSRV(t *testing.T) {
+	t.Parallel()
+
+	logger := blog.NewMock()
+	ring := newTestRedisRing()
+
+	_, err := NewLookup([]cmd.ServiceDomain{
+		{
+			Service: "doesnotexist",
+			Domain:  "service.consuls",
+		},
+		{
+			Service: "doesnotexist2",
+			Domain:  "service.consuls",
+		},
+	},
+		"consul.service.consul",
+		250*time.Millisecond,
+		ring,
+		logger,
+	)
+	test.AssertError(t, err, "Expected NewLookup construction to fail")
+}
+
+func TestLookupWithAllFailingSRV(t *testing.T) {
+	t.Parallel()
+
+	logger := blog.NewMock()
+	ring := newTestRedisRing()
+
+	lookup, err := NewLookup([]cmd.ServiceDomain{
+		{
+			Service: "redisratelimits",
+			Domain:  "service.consul",
+		},
+	},
+		"consul.service.consul",
+		250*time.Millisecond,
+		ring,
+		logger,
+	)
+	test.AssertNotError(t, err, "Expected NewLookup construction to succeed")
+
+	lookup.srvLookups = []cmd.ServiceDomain{
+		{
+			Service: "doesnotexist1",
+			Domain:  "service.consul",
+		},
+		{
+			Service: "doesnotexist2",
+			Domain:  "service.consul",
+		},
+	}
+
+	testCtx, cancel := context.WithCancel(context.Background())
+	tempErr, nonTempErr := lookup.updateNow(testCtx)
+	cancel()
+	test.AssertNotError(t, tempErr, "Expected no temporary errors")
+	test.AssertError(t, nonTempErr, "Expected non-temporary errors to have occurred")
+}
+
+func TestLookupWithAllFailingSRVs(t *testing.T) {
+	t.Parallel()
+
+	logger := blog.NewMock()
+	ring := newTestRedisRing()
+
+	lookup, err := NewLookup([]cmd.ServiceDomain{
+		{
+			Service: "redisratelimits",
+			Domain:  "service.consul",
+		},
+	},
+		"consul.service.consul",
+		250*time.Millisecond,
+		ring,
+		logger,
+	)
+	test.AssertNotError(t, err, "Expected NewLookup construction to succeed")
+
+	// Replace the dnsAuthority with a non-existent DNS server, this will cause
+	// a timeout error, which is technically a temporary error, but will
+	// eventually result in a non-temporary error when no shards are resolved.
+	lookup.dnsAuthority = "consuls.services.consuls:53"
+
+	testCtx, cancel := context.WithCancel(context.Background())
+	tempErr, nonTempErr := lookup.updateNow(testCtx)
+	cancel()
+	test.AssertError(t, tempErr, "Expected temporary errors")
+	test.AssertError(t, nonTempErr, "Expected a non-temporary error")
+	test.AssertErrorIs(t, nonTempErr, ErrNoShardsResolved)
+}
+
+func TestLookupWithOneFailingSRV(t *testing.T) {
+	t.Parallel()
+
+	logger := blog.NewMock()
+	ring := newTestRedisRing()
+
+	lookup, err := NewLookup([]cmd.ServiceDomain{
+		{
+			Service: "doesnotexist",
+			Domain:  "service.consuls",
+		},
+		{
+			Service: "redisratelimits",
+			Domain:  "service.consul",
+		},
+	},
+		"consul.service.consul",
+		250*time.Millisecond,
+		ring,
+		logger,
+	)
+	test.AssertNotError(t, err, "Expected NewLookup construction to succeed")
+
+	testCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	lookup.Start(testCtx)
```

Comment:
Is this call necessary? Now that NewLookup does a call to updateNow, and you force another call to updateNow on line 218, I don't think that actually starting the background goroutine is necessary for this test.

Reply:
You're correct, this isn't necessary. I've added this call to ensure that the codepath is exercised inside of the unit tests.

Reply:
I've gone ahead and moved this call to its own test case.